Aquarium/Engine/Tests/BoardTests.swift

```diff
@@ -0,0 +1,261 @@
+                     [v, v, v, v, v, a]]
+
+        var (ok, status): (Bool, Board.Status)
+
+        (ok, status) = board.validCols()
+        XCTAssertEqual(ok, false)
+        XCTAssertEqual(status, .columnTooMuchAir(2))
+
+        board.mat[1][2] = .void
+        board.mat[0][2] = .void
+
+        (ok, status) = board.validCols()
+        XCTAssertEqual(ok, false)
+        XCTAssertEqual(status, .columnTooMuchAir(3))
+    }
+}
+
+final class SolveStateTests: XCTestCase {
+    private let v = Cell.void, w = Cell.water, a = Cell.air
+
+    func testRowSolveState() throws {
+        var board = Board.empty(size: 6)
+        board.colSums = [2, 4, 5, 5, 4, 2]
+        board.rowSums = [4, 3, 4, 4, 2, 5]
+
+        board.mat = [[v, v, a, a, v, v],
+                     [w, w, a, a, w, w],
+                     [v, w, a, a, w, v],
+                     [a, a, w, w, a, a],
+                     [w, v, w, w, v, v],
+                     [v, v, v, v, v, a]]
+        XCTAssertFalse(board.allRowsSolved)
+
+        board.mat = [[a, w, w, w, w, a],
+                     [a, a, w, w, w, a],
+                     [a, w, w, w, w, a],
+                     [a, w, w, w, w, a],
+                     [w, a, a, a, a, w],
+                     [w, w, w, w, a, w]]
+
+        XCTAssertTrue(board.allRowsSolved)
+    }
+
+    func testAllColumnsSolved() throws {
+        var board = Board.empty(size: 6)
+        board.colSums = [2, 4, 5, 5, 4, 2]
+        board.rowSums = [4, 3, 4, 4, 2, 5]
+
+        board.mat = [[v, v, a, a, v, v],
+                     [w, w, a, a, w, w],
+                     [v, w, a, a, w, v],
+                     [a, a, w, w, a, a],
+                     [w, v, w, w, v, v],
+                     [v, v, v, v, v, a]]
+        XCTAssertFalse(board.allColsSolved)
+
+        board.mat = [[a, w, w, w, w, a],
+                     [a, a, w, w, w, a],
+                     [a, w, w, w, w, a],
+                     [a, w, w, w, w, a],
+                     [w, a, a, a, a, w],
+                     [w, w, w, w, a, w]]
+        XCTAssertTrue(board.allColsSolved)
+    }
+}
+
+final class FlowTests: XCTestCase {
+    /**
+     * One common board for all flow validity tests
+     */
+    private func board() -> Board {
+        let v = Cell.void, w = Cell.water, a = Cell.air
+        var board = Board(
+            colSums: [2, 4, 5, 5, 4, 2],
+            rowSums: [4, 3, 4, 4, 2, 5],
+            groups: [[1, 2, 2, 2, 2, 3],
+                     [1, 1, 2, 4, 2, 3],
+                     [1, 2, 2, 4, 4, 3],
+                     [1, 2, 2, 2, 2, 3],
+                     [1, 5, 5, 5, 6, 3],
+                     [1, 1, 1, 5, 6, 3]]
+        )
+        board.mat = [[v, v, a, a, v, w],
+                     [w, w, a, a, w, w],
+                     [v, w, a, a, w, v],
+                     [a, a, w, w, a, a],
+                     [w, v, w, w, v, v],
+                     [w, v, v, v, v, w]]
+        return board
+    }
```

Comment:
Maybe this can be a computed variable e.g. called `commonBoard` instead? Since it's a nullary function.

Pretty clean solution compared to my copy-paste tho HAHHA